### ISSUE TITLE
Refactor go-hep.org/x/hep/fit to gonum

### ISSUE THREAD

**Issue by fawick** [gonum/gonum#296]
I'd like to propose to refactor the [fit package from go-hep](godoc.org/go-hep.org/x/hep/fit) to gonum. I strongly believe that the function fitting would benefit more application domains besides High-Energy Physics, and I feel that the dependency on the go-hep project in those cases is to heavy.

**Comment by sbinet** [comment]
that's fine by me (and was the intention all along)
but I guess Gonum at large needs to decide :)

**Comment by kortschak** [comment]
You have a vote from me.

**Comment by sbinet** [comment]
ok. I'll prepare a PR with a new `gonum.org/v1/gonum/fit` package :)

**Comment by btracey** [comment]
A lot of people would like this to happen, but I think there should be some planning first about how. There are a lot of parts to a good fitting package, and the design isn't easy. I have a bunch of thoughts from various efforts at building one, but the broader point is seeing some discussion of interfaces and composability before jumping into code.

**Comment by sbinet** [comment]
oops, our mails crossed each others...

**Comment by sbinet** [comment]
I am happy for a discussion to happen :)

**Comment by fawick** [comment]
@btracey: I'd consider reviewing an actual PR to be much more convenient, even for a discussing the design, as one would be able to comment on functions/methods etc. directly. There is always the option not to merge a PR if the design isn't right.

**Comment by btracey** [comment]
In some cases I agree that that's true. However, the scope of a good `fit` package is much more compartmentalized and larger than the existing `hep/fit`. For instance, `hep/fit` only does least-squares, as opposed to other common objective functions (cross entropy, etc.), has nothing about regularization, assumes an `optimize.Method`, which means it can't do global optimizers or Stochastic Gradient Descent, assumes a fixed data set, meaning it can't do adaptation to online data, etc.

This is not to say `hep/fit` is bad, it's not, and I also have a couple of packages with smaller scope that serve useful functions even though they aren't right for the more general problem. I just think there's not a lot of harm in having useful routines live in other places for the time being, and there is a lot of potential upside to planning so `gonum` ends with a high quality package.
